### ISSUE TITLE
Use an ordered set of types in TypeUniverse

### ISSUE THREAD

**Issue by sysrqb** [mozilla/uniffi-rs#374]
Currently, `types::TypeUniverse:all_known_types` is a `HashSet`, unfortunately this breaks reproducible builds because uniffi_bindgen iterates over that set (in an undefined order) when generating bindings resulting in different ordering in the generated files (see https://gitlab.torproject.org/tpo/applications/tor-browser-build/-/issues/40208 for some context, starting at Fenix and working down to uniffi_bindgen :) )

As one possible solution, substituting `HashSet` for `BTreeSet`, and adding `PartialOrd` and `Ord` as derived traits seems to solve this problem.

┆Issue is synchronized with this [Jira Task](https://jira.mozilla.com/browse/SYNC-2014)

**Comment by sysrqb** [comment]
Specifically, this is the patch I'm currently testing on v0.5.0

```diff
diff --git a/uniffi_bindgen/src/interface/types.rs b/uniffi_bindgen/src/interface/types.rs
index 6448d58..b7efd22 100644
--- a/uniffi_bindgen/src/interface/types.rs
+++ b/uniffi_bindgen/src/interface/types.rs
@@ -26,7 +26,7 @@
 use anyhow::bail;
 use anyhow::Result;
 use std::convert::TryFrom;
-use std::{collections::hash_map::Entry, collections::HashMap, collections::HashSet};
+use std::{collections::hash_map::Entry, collections::HashMap, collections::BTreeSet};
 
 use super::Attributes;
 
@@ -71,7 +71,7 @@ pub enum FFIType {
 /// Represents all the different high-level types that can be used in a component interface.
 /// At this level we identify user-defined types by name, without knowing any details
 /// of their internal structure apart from what type of thing they are (record, enum, etc).
-#[derive(Debug, Clone, Eq, PartialEq, Hash)]
+#[derive(Debug, Clone, Eq, PartialEq, Hash, PartialOrd, Ord)]
 pub enum Type {
     // Primitive types.
     UInt8,
@@ -195,7 +195,7 @@ pub(crate) struct TypeUniverse {
     // Named type definitions (including aliases).
     type_definitions: HashMap<String, Type>,
     // All the types in the universe, by canonical type name.
-    all_known_types: HashSet<Type>,
+    all_known_types: BTreeSet<Type>,
 }
 
 impl TypeUniverse {
```

**Comment by mhammond** [comment]
That sounds reasonable, especially with a few comments indicating why ordered sets are important. If that ends up solving your problem, please put up a PR.

**Comment by rfk** [comment]
Thanks for investigating, and as Mark says above, I think we'd be very open to taking this change.

**Comment by sysrqb** [comment]
Thanks @mhammond and @rfk. I'm finally getting back to this. I opened #433

**Comment by mhammond** [comment]
Fixed by #433
